crates/sui-graphql-rpc/src/types/transaction_block.rs

```diff
@@ -297,102 +305,212 @@ impl TransactionBlock {
-                        }
+        use checkpoints::dsl as cp;
+        use transactions::dsl as tx;
 
-                        if let Some(txs) = &filter.transaction_ids {
-                            let digests: Vec<_> = txs.iter().map(|d| d.to_vec()).collect();
-                            query = query.filter(tx::dsl::transaction_digest.eq_any(digests));
+        let (prev, next, transactions): (bool, bool, Vec<StoredTransaction>) = db
+            .execute_repeatable(move |conn| {
+
+                // The min or first `tx_sequence_number` of a checkpoint is the previous
+                // checkpoint's `network_total_transactions`. Because this refers to a historical
+                // checkpoint, if we yield a `None` result, we can return early.
+                let lo_tx = match lo_cp {
+                    0 => 0,
+                    _ => {
+                        let sequence_number: Option<i64> = conn
+                        .first(move || {
+                            cp::checkpoints
+                                .select(cp::network_total_transactions)
+                                .filter(cp::sequence_number.eq((lo_cp - 1) as i64))
+                        }).optional()?;
+
+                        match sequence_number {
+                            Some(sequence_number) => sequence_number as u64,
+                            None => return Ok::<_, diesel::result::Error>((false, false, vec![])),
                         }
-
-                        query
-                    },
-                )
+                    }
+                };
+
+                // The max or last `tx_sequence_number` of a checkpoint is the current checkpoint's
+                // `network_total_transactions` - 1. Since we cap the upper bound to
+                // `checkpoint_viewed_at`, we should always yield a result; thus, we can also return
+                // early if result is `None`.
+                let hi_tx = {
+                    let sequence_number: Option<i64> = conn
+                        .first(move || {
+                            cp::checkpoints
+                                .select(cp::network_total_transactions)
+                                .filter(cp::sequence_number.eq(hi_cp as i64))
+                        }).optional()?;
+
+                    match sequence_number {
+                        Some(sequence_number) => (sequence_number as u64).saturating_sub(1),
+                        None => return Ok::<_, diesel::result::Error>((false, false, vec![])),
+                    }
+                };
+
+                let sender = filter.sign_address;
+
+                let mut bound = TxLookupBound::from_range((Some(lo_tx), Some(hi_tx)));
+
+                // If `transaction_ids` is specified, we can use that in lieu of the range under the
+                // assumption that it will further constrain the rows we look up.
+                if let Some(txs) = &filter.transaction_ids {
```

Comment:
Couple of things we need to do in the case where transaction IDs are provided:

- We still need to apply any transaction bounds (for pagination purposes).
- We need to enforce a limit on the number of transaction IDs are allowed in a filter. Originally, I was thinking the same as the max page size limit, but we can probably afford to be an order of magnitude more generous.
- Like the case where we don't have transaction digests, the query is split into two parts: fetch sequence numbers, and then multi-get from `transactions`, but when we have transaction digests we need to apply additional filters on the `transactions` table query itself, to avoid the query planner making bad choices with joins (this will force the query to be an index lookup by primary key with an additional filter).